specs/governance/alligator.md

```diff
@@ -0,0 +1,207 @@
+
+The `Alligator` contract migrates the delegation state from the [`GovernanceToken`](gov-token.md) to itself
+through a hook-based approach. Specifically, the [`GovernanceToken`](gov-token.md) calls the `Alligator` contract's
+`afterTokenTransfer` function after a token transfer. This enables the `Alligator` contract to consume the hook and update
+its delegation and checkpoint mappings accordingly. If either address involved in the transfer (`_from_` or `_to`) has not
+been migrated to the `Alligator` contract, the contract copies the address' checkpoint data from the
+[`GovernanceToken`](gov-token.md) to its own state.
+
+## Interface
+
+### Core Functions
+
+#### `subdelegate`
+
+Allows subdelegation of token voting power to another address (delegatee) with specified subdelegation rules. This function
+is inteded to be called by users that require advanced delegation of the [`GovernanceToken`](gov-token.md).
+
+```solidity
+subdelegate(address _delegatee, SubdelegationRule _rule)
+```
+
+#### `subdelegateFromToken`
+
+Allows subdelegation of token voting power from an address (delegator) to another address (delegatee) with a specified
+subdelegation rule. This function is intended to be called by the [`GovernanceToken`](gov-token.md) contract as part
+of its `delegateBySig` function. To ensure backwards compatibility in the [`GovernanceToken`](gov-token.md), the
+subdelegation rule must be 100% delegation, mimicking the behavior of the [`GovernanceToken`](gov-token.md)'s `delegate`
+function.
+
+
+```solidity
+subdelegateFromToken(address _delegator, address _delegatee, SubdelegationRule _rule)
+```
+
+#### `subdelegateBatched`
+
+Allows batch subdelegation of token voting power to multiple addresses with specified subdelegation rules. This
+function is intended to be called by users.
+
+```solidity
+subdelegateBatched(address[] _delegatees, SubdelegationRule[] _rules)
+```
+
+Calls the `subdelegate` function for each pair of `_delegatees` address and subdelegation rule.
+
+#### `afterTokenTransfer`
+
+Updates the voting power of two addresses (`_from` and `_to`) after a token transfer. This function MUST
+be called by the `_afterTokenTransfer` function in the [`GovernanceToken`](gov-token.md).
+
+```solidity
+afterTokenTransfer(address _from, address _to, uint256 _amount)
+```
+
+The `Alligator` MUST check if the `_from` or `_to` addresses have been migrated by checking the `migrated` mapping
+from its [storage](#storage). If either address has not been migrated, the `Alligator` MUST copy the delegation
+and checkpoint data from the token contract to its own state. After copying the data, the `Alligator` MUST update
+the `migrated` mapping to reflect that the address has been migrated.
+
+### Getters
+
+For backwards compatibility, the `Alligator` MUST implement all public getter functios of the
+[`GovernanceToken`](gov-token.md) related to delegation and voting power. These functions MUST be used by the
```

Comment:
Do each of these getters check the `migrated` mapping and then forward to `GovToken` if not migrated?

Reply:
I think we should do on Alligator as well. Otherwise, the voting power / delegation state reflected on the `Alligator` won't be accurate when an address hasn't been `migrated`

Reply:
agreed. updated in https://github.com/ethereum-optimism/specs/pull/247/commits/7752d7d2abc4040f3cbd4104ace81be69067bbf1